app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+        } catch (e : FileNotFoundException) {
+            throw FileNotFoundException("Couldn't move document $sourceDocumentId")
+        }
+    }
+
+    private fun includeFile(result : MatrixCursor, documentId : String?, file : File?) : MatrixCursor {
+        var localDocumentId = documentId
+        var localFile = file
+
+        if (localDocumentId == null) {
+            localDocumentId = localFile?.let { getDocIdForFile(it) } as String?
+        } else if (localFile == null) {
+            localFile = getFileForDocId(localDocumentId)
+        }
+
+        var flags = 0
+        if (localFile == null) {
+            throw FileNotFoundException()
+        } else {
+            if (localFile.isDirectory) {
+                if (localFile.isDirectory && localFile.canWrite()) {
+                    flags = flags or DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE
+                }
+            } else if (localFile.canWrite()) {
+                flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_WRITE
+                flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_DELETE
+
+                flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_REMOVE
+                flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_MOVE
+                flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_COPY
+                flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_RENAME
+            }
+
+            val displayName = localFile.name
+            val mimeType = getTypeForFile(localFile)
+
+            result.newRow().apply {
+                add(DocumentsContract.Document.COLUMN_DOCUMENT_ID, localDocumentId)
+                add(DocumentsContract.Document.COLUMN_DISPLAY_NAME, displayName)
+                add(DocumentsContract.Document.COLUMN_SIZE, localFile.length())
+                add(DocumentsContract.Document.COLUMN_MIME_TYPE, mimeType)
+                add(DocumentsContract.Document.COLUMN_LAST_MODIFIED, localFile.lastModified())
+                add(DocumentsContract.Document.COLUMN_FLAGS, flags)
+            }
+        }
+        return result
+    }
+
+    private fun getTypeForFile(file : File) : Any? {
+        return if (file.isDirectory) {
+            DocumentsContract.Document.MIME_TYPE_DIR
+        } else {
+            getTypeForName(file.name)
+        }
+    }
+
+    private fun getTypeForName(name : String) : Any? {
+        val lastDot = name.lastIndexOf('.')
+        if (lastDot >= 0) {
+            val extension = name.substring(lastDot + 1)
+            val mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
+            if (mime != null) {
+                return mime
+            }
+        }
+        return "application/octect-stream"
+    }
+
+    private fun getFileForDocId(documentId : String) : File {
+        val f = File(documentId)
+        if (!f.exists()) throw FileNotFoundException(f.absolutePath + " not found")
+        return f
+    }
```

Comment:
You'll want to move this right up to where `getDocumentId` is and rename it to `getFile` and add a comment `@return The [File] that corresponds to the document ID supplied by [getDocumentId]`. Additionally, expand `f` to `file` as that's more consistent with the rest of the file and use string templates for the exception.